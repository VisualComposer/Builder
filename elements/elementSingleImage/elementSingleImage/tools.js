import { getService } from 'vc-cake'
const { parseDynamicBlock } = getService('utils')

const parseSize = (size, isRound, naturalWidth, naturalHeight) => {
  let crop = true
  if (typeof size === 'string') {
    size = size.replace(/\s/g, '').replace(/px/g, '').toLowerCase().split('x')
  } else if (typeof size === 'object') {
    crop = size.crop
    size = [size.width, size.height]
  }

  naturalWidth = parseInt(naturalWidth)
  naturalHeight = parseInt(naturalHeight)

  const cropHorizontal = parseInt(size[0]) < naturalWidth
  const cropVertical = parseInt(size[1]) < naturalHeight

  if (crop) {
    size[0] = parseInt(size[0]) < naturalWidth ? parseInt(size[0]) : naturalWidth
    size[1] = parseInt(size[1]) < naturalHeight ? parseInt(size[1]) : naturalHeight
  } else {
    size[0] = cropHorizontal ? parseInt(size[0]) : naturalWidth
    size[1] = cropVertical ? parseInt(size[1]) : naturalHeight

    if (cropHorizontal && !cropVertical) {
      const prop = size[0] / naturalWidth
      size[1] = parseInt(naturalHeight * prop)
    }

    if (cropVertical && !cropHorizontal) {
      const prop = size[1] / naturalHeight
      size[0] = parseInt(naturalWidth * prop)
    }

    if (cropVertical && cropHorizontal) {
      if (naturalHeight < naturalWidth) {
        const prop = size[0] / naturalWidth
        size[1] = parseInt(naturalHeight * prop)
      } else {
        const prop = size[1] / naturalHeight
        size[0] = parseInt(naturalWidth * prop)
      }
    }
  }

  if (isRound) {
    const smallestSize = size[0] >= size[1] ? size[1] : size[0]
    size = {
      width: smallestSize,
      height: smallestSize
    }
  } else {
    size = {
      width: size[0],
      height: size[1]
    }
  }
  return size
}

const checkRelatedSize = (size) => {
  let relatedSize = null
  if (window.vcvImageSizes && window.vcvImageSizes[size]) {
    relatedSize = window.vcvImageSizes[size]
  }
  return relatedSize
}

export const getSizes = (atts, img) => {
  if (!img) {
    return {
      width: '',
      height: ''
    }
  }
  let { size, shape } = atts
  size = size.replace(/\s/g, '').replace(/px/g, '').toLowerCase()

  let parsedSize = ''

  if (size.match(/\d+(x)\d+/)) {
    parsedSize = parseSize(size, shape === 'round', img.width, img.height)
  } else {
    parsedSize = checkRelatedSize(size)

    if (parsedSize) {
      parsedSize = parseSize(parsedSize, shape === 'round', img.width, img.height)
    } else {
      parsedSize = parseSize({ width: img.width, height: img.height }, shape === 'round', img.width, img.height)
    }
  }

  return {
    width: parsedSize.width,
    height: parsedSize.height
  }
}

export const getImageShortcode = (options, image, state) => {
  const { props, classes, isDefaultImage, src, isDynamicImage, naturalSizes } = options

  let shortcode = `[vcvSingleImage class="${classes}" data-width="${state.parsedWidth || 0}" data-height="${state.parsedHeight || 0}" src="${src}" data-img-src="${props['data-img-src']}" data-attachment-id="${props['data-attachment-id']}" `

  let alt = props.alt
  let title = props.title
  if (isDefaultImage) {
    shortcode += ' data-default-image="true"'
  }

  if (isDynamicImage) {
    const blockInfo = parseDynamicBlock(image.full)
    shortcode += ` data-dynamic="${blockInfo.blockAtts.value}"`
    try {
      const url = new URL(blockInfo.blockAtts.currentValue)
      const urlParams = new URLSearchParams(url.search)

      if (urlParams.get('alt')) {
        alt = urlParams.get('alt')
      }

      if (urlParams.get('title')) {
        title = urlParams.get('title')
      }
    } catch (e) {
      // In case if dynamicField (noValue)
      console.warn('URL is not valid, skipping', blockInfo, blockInfo.blockAtts.currentValue)
    }

    if (naturalSizes) {
      shortcode += ' data-dynamic-natural-size="true"'
    }
  }

  shortcode += ` alt="${alt}" title="${title}" ]`

  return shortcode
}
