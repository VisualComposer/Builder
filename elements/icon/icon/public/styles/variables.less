// icon sizes
@vce-icon-font-size-tiny: 18px;
@vce-icon-font-size-small: 24px;
@vce-icon-font-size-medium: 32px;
@vce-icon-font-size-large: 48px;
