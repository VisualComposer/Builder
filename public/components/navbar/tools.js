/**
 * Get closest parent to element
 * @param el
 * @param selector
 * @returns {*}
 */
export const getClosest = (el, selector) => {
  let matchesFn;
  // find vendor prefix
  ['matches', 'webkitMatchesSelector', 'mozMatchesSelector', 'msMatchesSelector', 'oMatchesSelector'].some(function (fn) {
    if (typeof document.body[fn] === 'function') {
      matchesFn = fn
      return true
    }
    return false
  })
  let parent
  // traverse parents
  while (el) {
    parent = el.parentElement
    if (parent && parent[matchesFn](selector)) {
      return parent
    }
    el = parent
  }
  return null
}

/**
 * Get real width and height for element in DOM.
 *
 * @param $el
 * @param container
 * @returns {*|NavbarControl.state.realSize|{width, height}}
 */
export const getRealSize = ($el, parent) => {
  const realSize = { width: 0, height: 0 }
  const $tempEl = $el.cloneNode(true)
  $tempEl.style.position = 'fixed'
  let container
  if ($el.closest === undefined) {
    container = getClosest($el, parent)
  } else {
    container = $el.closest(parent)
  }
  container.appendChild($tempEl)
  const style = window.getComputedStyle($tempEl, null)
  // get width
  realSize.width = $tempEl.offsetWidth
  realSize.width += parseInt(style.marginLeft) + parseInt(style.marginRight)
  // get height
  realSize.height = $tempEl.offsetHeight
  realSize.height += parseInt(style.marginLeft) + parseInt(style.marginRight)

  $tempEl.remove()
  return realSize
}
