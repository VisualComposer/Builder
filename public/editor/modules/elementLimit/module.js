import { getStorage, getService } from 'vc-cake'
import store from 'public/editor/stores/store'
import { notificationAdded } from 'public/editor/stores/notifications/slice'

const elementsStorage = getStorage('elements')
const documentManager = getService('document')
const cook = getService('cook')
const dataManager = getService('dataManager')

const localizations = dataManager.get('localizations')
const limitTexts = {
  elementLimit1: localizations.onlyOneElementCanBeAddedToPage || 'Only one %element element can be added to the page.',
  elementLimit2: localizations.onlyTwoElementsCanBeAddedToPage || 'Only two %element elements can be added to the page.',
  elementLimit3: localizations.onlyThreeElementsCanBeAddedToPage || 'Only three %element elements can be added to the page.',
  elementLimit4: localizations.onlyFourElementsCanBeAddedToPage || 'Only four %element elements can be added to the page.',
  elementLimit5: localizations.onlyFiveElementsCanBeAddedToPage || 'Only five %element elements can be added to the page.',
  elementLimitDefault: localizations.elementLimitDefaultText || 'Only %count %element elements can be added to the page.'
}

const triggerNotification = (type, elementName, limit) => {
  let limitText = limitTexts.elementLimitDefault
  if (limit < 6) {
    limitText = limitTexts[`elementLimit${limit}`]
  } else {
    limitText = limitText.replace('%count', limit)
  }
  limitText = limitText.replace('%element', elementName)

  store.dispatch(notificationAdded({
    type: type,
    text: limitText,
    time: 5000,
    showCloseButton: true
  }))
}

const getElementExceededLimitStatus = (element) => {
  const limitData = {}
  if (Object.prototype.hasOwnProperty.call(element, 'metaElementLimit')) {
    const limit = parseInt(element.metaElementLimit)
    const limitedElements = documentManager.getByTag(element.tag) || {}
    if (limit > 0 && Object.keys(limitedElements).length >= limit) {
      limitData.hasExceeded = true
      limitData.limit = limit
    }
  }
  return limitData
}

elementsStorage.registerAction('beforeAdd', (element) => {
  const elementLimitData = getElementExceededLimitStatus(element)
  if (elementLimitData.hasExceeded) {
    const cookElement = cook.get(element)
    triggerNotification('error', cookElement.get('name'), elementLimitData.limit)
    return true
  }
  return false
})

elementsStorage.registerAction('beforeClone', (elementID) => {
  const element = documentManager.get(elementID)
  const elementLimitData = getElementExceededLimitStatus(element)
  if (elementLimitData.hasExceeded) {
    const cookElement = cook.get(element)
    triggerNotification('error', cookElement.get('name'), elementLimitData.limit)
    return true
  }
  return false
})
