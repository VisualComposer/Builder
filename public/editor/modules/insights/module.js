import { getStorage, getService, env, add } from 'vc-cake'
import { debounce, memoize } from 'lodash'

const insightsStorage = getStorage('insights')
const historyStorage = getStorage('history')
const settingsStorage = getStorage('settings')
const workspaceStorage = getStorage('workspace')
const elementsStorage = getStorage('elements')
const cookService = getService('cook')
const utils = getService('utils')
const dataManager = getService('dataManager')

add('insights', () => {
  // VC: Insights
  class InsightsChecks {
    isImagesSizeLarge = false
    localizations = dataManager.get('localizations')

    checkTitleLength () {
      if (dataManager.get('editorType') !== 'default') {
        return
      }

      let pageTitleLength = settingsStorage.state('pageTitle').get().length
      if (settingsStorage.state('pageTitleDisabled').get()) {
        const headings = env('iframe').document.body.querySelectorAll('h1')
        for (let i = 0; i < headings.length; i++) {
          const heading = headings[i]
          if (heading.offsetParent !== null && heading.getBoundingClientRect().height) {
            pageTitleLength = heading.innerText.length
            break
          }
        }
      }

      const insightsTitleTooLong = this.localizations.insightsTitleTooLong
      const insightsTitleTooLong60 = this.localizations.insightsTitleTooLong60
      const insightsTitleTooLong100 = this.localizations.insightsTitleTooLong100
      const insightsTitleTooShort = this.localizations.insightsTitleTooShort
      const insightsTitleTooShortDescription = this.localizations.insightsTitleTooShortDescription
      const insightsTitleGood = this.localizations.insightsTitleGood

      if (pageTitleLength > 100) {
        insightsStorage.trigger('add', {
          state: 'critical',
          type: 'titleLength',
          title: insightsTitleTooLong,
          groupDescription: insightsTitleTooLong100
        })
      } else if (pageTitleLength > 61) {
        insightsStorage.trigger('add', {
          state: 'warning',
          type: 'titleLength',
          title: insightsTitleTooLong,
          groupDescription: insightsTitleTooLong60
        })
      } else if (pageTitleLength > 9) {
        insightsStorage.trigger('add', {
          state: 'success',
          type: 'titleLength',
          title: insightsTitleGood,
          groupDescription: `Your page title consists of ${pageTitleLength} characters which is optimal title size.`
        })
      } else if (pageTitleLength >= 0) {
        insightsStorage.trigger('add', {
          state: 'warning',
          type: 'titleLength',
          title: insightsTitleTooShort,
          groupDescription: insightsTitleTooShortDescription
        })
      }
    }

    checkNoIndex () {
      const noIndexMetaTag = this.localizations.noIndexMetaTag
      const noIndexMetaTagDescription = this.localizations.noIndexMetaTagDescription

      const metas = document.querySelector('.vcv-layout-iframe').contentWindow.document.querySelectorAll('meta')
      metas.forEach((meta) => {
        if (meta.content.includes('noindex')) {
          insightsStorage.trigger('add', {
            state: 'warning',
            type: 'noIndex',
            title: noIndexMetaTag,
            groupDescription: noIndexMetaTagDescription
          })
        }
      })
    }

    checkLinks () {
      const inboundLinks = []
      const outboundLinks = []

      const noInboundLinks = this.localizations.noInboundLinks
      const noInboundLinksDescription = this.localizations.noInboundLinksDescription
      const noOutboundLinks = this.localizations.noOutboundLinks
      const noOutboundLinksDescription = this.localizations.noOutboundLinksDescription

      const contentRoot = env('iframe').document.getElementById('vcv-editor')
      const links = contentRoot.querySelectorAll('a')
      links.forEach((link) => {
        if (window.location.host === link.host) {
          inboundLinks.push(link)
        } else {
          outboundLinks.push(link)
        }
      })

      if (!inboundLinks.length) {
        insightsStorage.trigger('add', {
          state: 'warning',
          type: 'noInboundLinks',
          title: noInboundLinks,
          groupDescription: noInboundLinksDescription
        })
      }

      if (!outboundLinks.length) {
        insightsStorage.trigger('add', {
          state: 'warning',
          type: 'noOutboundLinks',
          title: noOutboundLinks,
          groupDescription: noOutboundLinksDescription
        })
      }
    }

    checkForHeadings () {
      if (dataManager.get('editorType') !== 'default') {
        return
      }
      const headings = env('iframe').document.body.querySelectorAll('h1')
      let visibleHeadings = 0
      for (let i = 0; i < headings.length; i++) {
        const heading = headings[i]
        if (heading.offsetParent !== null && heading.getBoundingClientRect().height) {
          visibleHeadings++
        }
      }
      if (visibleHeadings === 0) {
        const h1MissingTitle = this.localizations.insightsH1MissingTitle
        const h1MissingDescription = this.localizations.insightsH1MissingDescription
        insightsStorage.trigger('add', {
          state: 'critical',
          type: 'noH1',
          title: h1MissingTitle,
          groupDescription: h1MissingDescription
        })
      } else if (visibleHeadings === 1) {
        const insightsH1ExistsTitle = this.localizations.insightsH1ExistsTitle
        const insightsH1ExistsDescription = this.localizations.insightsH1ExistsDescription
        insightsStorage.trigger('add', {
          state: 'success',
          type: 'existH1',
          title: insightsH1ExistsTitle,
          groupDescription: insightsH1ExistsDescription
        })
      } else {
        const insightsMultipleH1Title = this.localizations.insightsMultipleH1Title
        const insightsMultipleH1Description = this.localizations.insightsMultipleH1Description
        insightsStorage.trigger('add', {
          state: 'critical',
          type: 'multipleH1',
          title: insightsMultipleH1Title,
          groupDescription: insightsMultipleH1Description
        })
      }
    }

    checkForEmptyContent () {
      const elements = getStorage('elements').state('document').get() || []
      if (!elements.length) {
        // There are no elements on a page
        insightsStorage.trigger('add', {
          state: 'critical',
          type: 'noElementsOnPage',
          title: this.localizations.insightsNoContentOnPageTitle,
          groupDescription: this.localizations.insightsNoContentOnPageDescription
        })
        return true
      }
      return false
    }

    checkForAlt () {
      const images = env('iframe').document.body.querySelectorAll('img')
      let allImagesHasAlt = true
      images.forEach((image) => {
        if (!image.alt || image.alt === '') {
          const altMissingTitle = this.localizations.insightsImageAltAttributeMissingTitle
          const description = this.localizations.insightsImageAltAttributeMissingDescription
          const elementId = InsightsChecks.getElementId(image)
          const position = InsightsChecks.getNodePosition(image)
          allImagesHasAlt = false
          insightsStorage.trigger('add', {
            state: 'critical',
            type: `altMissing${position}`,
            thumbnail: image.src,
            title: position !== 'Content' ? `${position}: ${altMissingTitle}` : altMissingTitle,
            groupDescription: description,
            description: 'Alt attribute is empty %s'.replace('%s', elementId ? `(${cookService.getById(elementId).getName()})` : '').trim(),
            elementID: elementId,
            domNode: image
          })
        }
      })
      if (images.length && allImagesHasAlt) {
        const altExistsTitle = this.localizations.insightsImageAltAttributeExistsTitle
        const altExistsDescription = this.localizations.insightsImageAltAttributeExistsDescription
        insightsStorage.trigger('add', {
          state: 'success',
          type: 'altExists',
          title: altExistsTitle,
          groupDescription: altExistsDescription
        })
      }
    }

    checkForImageSize () {
      const images = env('iframe').document.body.querySelectorAll('img')
      const promises = []
      images.forEach((image) => {
        // Skip checking for our images
        if (image.src.indexOf('cdn.hub.visualcomposer.com') === -1) {
          promises.push(this.getImageSize(image.src, image))
        }
      })
      return promises
    }

    checkForBgImageSize () {
      function getBgImgs (doc) {
        const srcChecker = /url\(\s*?['"]?\s*?(\S+?)\s*?["']?\s*?\)/i
        return Array.from(
          // Remove .mce-tinymce, script, style from checking bg image
          Array.from(doc.querySelectorAll('body > *:not(.mce-tinymce, script, style) *:not(script, style), body')).reduce((collection, node) => {
            const prop = window.getComputedStyle(node, null)
              .getPropertyValue('background-image')
            const match = srcChecker.exec(prop)
            if (match) {
              collection.add({ src: match[1], domNode: node })
            }
            return collection
          }, new Set())
        )
      }

      const bgImages = getBgImgs(env('iframe').document)
      const promises = []
      bgImages.forEach((data) => {
        promises.push(this.getImageSize(data.src, data.domNode, 'background'))
      })
      return promises
    }

    async checkForImagesSize () {
      const promises = this.checkForImageSize()
      promises.concat(this.checkForBgImageSize())
      await Promise.all(promises)

      if (promises.length && !this.isImagesSizeLarge) {
        const imageSizeProperTitle = this.localizations.insightsImagesSizeProperTitle
        const imageSizeProperDescription = this.localizations.insightsImagesSizeProperDescription
        insightsStorage.trigger('add', {
          state: 'success',
          type: 'imgSizeProper',
          title: imageSizeProperTitle,
          groupDescription: imageSizeProperDescription
        })
      }
    }

    async getImageSize (src, domNode, type = '') {
      const imageSizeBytes = await InsightsChecks.getImageSizeRequest(src)
      if (imageSizeBytes && imageSizeBytes >= 1024 * 1024) {
        const imageSizeBigTitle = type === 'background' ? this.localizations.insightsBgImageSizeBigTitle : this.localizations.insightsImageSizeBigTitle
        let description = this.localizations.insightsImageSizeBigDescription
        const position = InsightsChecks.getNodePosition(domNode)
        const elementId = InsightsChecks.getElementId(domNode)
        const imageSizeInMB = imageSizeBytes / 1024 / 1024
        description = description.replace('%s', '1 MB')
        this.isImagesSizeLarge = true
        insightsStorage.trigger('add', {
          state: 'critical',
          type: `imgSize1MB${position}`,
          thumbnail: src,
          title: position !== 'Content' ? `${position}: ${imageSizeBigTitle}` : imageSizeBigTitle,
          groupDescription: description,
          description: 'Image size is' + ` ${imageSizeInMB.toFixed(2)} MB`,
          elementID: elementId,
          domNode: domNode
        })
      } else if (imageSizeBytes && imageSizeBytes >= 500 * 1024) {
        const imageSizeBigTitle = type === 'background' ? this.localizations.insightsBgImageSizeBigTitle : this.localizations.insightsImageSizeBigTitle
        let description = this.localizations.insightsImageSizeBigDescription
        const position = InsightsChecks.getNodePosition(domNode)
        const elementId = InsightsChecks.getElementId(domNode)
        description = description.replace('%s', '500 KB')
        this.isImagesSizeLarge = true
        insightsStorage.trigger('add', {
          state: 'warning',
          type: `imgSize500KB${position}`,
          thumbnail: src,
          title: position !== 'Content' ? `${position}: ${imageSizeBigTitle}` : imageSizeBigTitle,
          groupDescription: description,
          description: 'Image size is' + ` ${parseInt(imageSizeBytes / 1024)} KB`,
          elementID: elementId,
          domNode: domNode
        })
      }
    }

    static getImageSizeRequest = memoize((imageUrl) => {
      return new Promise((resolve, reject) => {
        const xhr = new window.XMLHttpRequest()
        xhr.open('HEAD', imageUrl, true)
        xhr.onload = function () {
          if (!(this.status >= 200 && this.status < 300)) {
            reject(new Error(`Wrong network status received: ${this.status} ${imageUrl}`))
          }
        }
        xhr.onreadystatechange = function () {
          if (xhr.readyState === 4 && xhr.status === 200) {
            resolve(xhr.getResponseHeader('Content-Length'))
          }
        }
        xhr.onerror = function (error) {
          reject(new Error(`Wrong network response received:${error}`))
        }
        xhr.send(null)
      })
    })

    static getElementId (domNode) {
      if (domNode.hasAttribute('data-vcv-element')) {
        return domNode.getAttribute('data-vcv-element')
      } else {
        const closestParent = domNode.closest('[data-vcv-element]')
        return closestParent ? closestParent.getAttribute('data-vcv-element') : false
      }
    }

    static getNodePosition (domNode) {
      const contentRoot = env('iframe').document.getElementById('vcv-editor')
      const documentPosition = domNode.compareDocumentPosition(contentRoot)
      if (documentPosition & window.Node.DOCUMENT_POSITION_CONTAINS) {
        return 'Content'
      } else if (documentPosition & window.Node.DOCUMENT_POSITION_FOLLOWING) {
        // Left Sidebar
        return domNode.closest('[data-vcv-layout-zone="sidebar"]') ? 'Left Sidebar' : 'Header'
      } else if (documentPosition & window.Node.DOCUMENT_POSITION_PRECEDING) {
        // Right Sidebar
        return domNode.closest('[data-vcv-layout-zone="sidebar"]') ? 'Right Sidebar' : 'Footer'
      }
    }

    checkParagraphsLength () {
      const paragraphs = env('iframe').document.body.querySelectorAll('.vcv-layouts-html p')
      let isParagraphSizeLarge = false

      paragraphs.forEach((paragraph) => {
        const paragraphLength = paragraph.innerText.split(' ').length
        if (paragraph.length && paragraphLength > 200) {
          const insightsParagraphLengthTitle = this.localizations.insightsParagraphLengthTitle
          const groupDescription = this.localizations.insightsParagraphLengthDescription200
          const description = this.localizations.insightsParagraphLengthDescription
          const elementId = InsightsChecks.getElementId(paragraph)
          const position = InsightsChecks.getNodePosition(paragraph)
          isParagraphSizeLarge = true
          insightsStorage.trigger('add', {
            state: 'critical',
            type: `paragraphLengthLarge${position}`,
            title: position !== 'Content' ? `${position}: ${insightsParagraphLengthTitle}` : insightsParagraphLengthTitle,
            groupDescription: groupDescription,
            description: `${description} ${paragraphLength}`,
            elementID: elementId,
            domNode: paragraph
          })
        } else if (paragraph.length && paragraphLength > 150 && paragraphLength < 200) {
          const elementId = InsightsChecks.getElementId(paragraph)
          const position = InsightsChecks.getNodePosition(paragraph)
          const insightsParagraphLengthTitle = this.localizations.insightsParagraphLengthTitle
          const groupDescription = this.localizations.insightsParagraphLengthDescription150
          const description = this.localizations.insightsParagraphLengthDescription
          isParagraphSizeLarge = true
          insightsStorage.trigger('add', {
            state: 'warning',
            type: `paragraphLengthMedium${position}`,
            title: position !== 'Content' ? `${position}: ${insightsParagraphLengthTitle}` : insightsParagraphLengthTitle,
            groupDescription: groupDescription,
            description: `${description} ${paragraphLength}`,
            elementID: elementId,
            domNode: paragraph
          })
        }
      })

      if (paragraphs.length && !isParagraphSizeLarge) {
        const insightsParagraphLengthTitle = this.localizations.insightsParagraphLengthTitle
        const groupDescription = this.localizations.insightsParagraphLengthDescriptionOk
        insightsStorage.trigger('add', {
          state: 'success',
          type: 'paragraphSizeProper',
          title: insightsParagraphLengthTitle,
          groupDescription: groupDescription
        })
      }
    }

    checkPostContentLength () {
      const elements = getStorage('elements').state('document').get() || []
      if (elements.length) {
        const layoutHTML = env('iframe').document.body.querySelector('.vcv-layouts-html').innerHTML
        const contentLength = utils.getTextContent(layoutHTML).split(/\s+/).length
        const insightsParagraphLengthTitle = this.localizations.insightsContentLengthTitle
        let groupDescription = ''

        if (layoutHTML && contentLength < 300) {
          groupDescription = this.localizations.insightsContentLengthDescription300
          insightsStorage.trigger('add', {
            state: 'critical',
            type: 'contentSize300',
            title: insightsParagraphLengthTitle,
            groupDescription: groupDescription.replace('%length', contentLength)
          })
        } else if (layoutHTML) {
          groupDescription = this.localizations.insightsContentLengthDescriptionOk
          insightsStorage.trigger('add', {
            state: 'success',
            type: 'contentSizeProper',
            title: insightsParagraphLengthTitle,
            groupDescription: groupDescription.replace('%length', contentLength)
          })
        }
      }
    }

    checkForGA () {
      const gaNodes = env('iframe').document.querySelectorAll('script[src*="googletagmanager.com"], script[src*="google-analytics.com"]')
      if (!window.dataLayer && !window.GoogleAnalyticsObject && !window.ga && !gaNodes.length) {
        const insightsGAMissingTitle = this.localizations.insightsGAMissingTitle
        const insightsGAMissingDescription = this.localizations.insightsGAMissingDescription
        insightsStorage.trigger('add', {
          state: 'warning',
          type: 'googleAnalytics',
          title: insightsGAMissingTitle,
          groupDescription: insightsGAMissingDescription
        })
      } else {
        const insightsGAMissingTitleOK = this.localizations.insightsGAMissingTitleOK
        const insightsGAMissingDescriptionOK = this.localizations.insightsGAMissingDescriptionOK
        insightsStorage.trigger('add', {
          state: 'success',
          type: 'googleAnalytics',
          title: insightsGAMissingTitleOK,
          groupDescription: insightsGAMissingDescriptionOK
        })
      }
    }
  }

  if (env('VCV_FT_INSIGHTS')) {
    const insightsStorageInstance = new InsightsChecks()
    const runChecksCallback = debounce(() => {
      if (!document.querySelector('.vcv-layout-iframe') || !document.querySelector('.vcv-layout-iframe').contentWindow) {
        return // editor reload
      }
      // clear previous <Insights>
      insightsStorage.trigger('reset')
      insightsStorageInstance.isImagesSizeLarge = false

      // Do all checks
      const isEmptyContent = insightsStorageInstance.checkForEmptyContent()
      if (!isEmptyContent) {
        insightsStorageInstance.checkForHeadings()
        insightsStorageInstance.checkForAlt()
        insightsStorageInstance.checkForImagesSize()
        insightsStorageInstance.checkParagraphsLength()
        insightsStorageInstance.checkTitleLength()
        insightsStorageInstance.checkNoIndex()
        insightsStorageInstance.checkPostContentLength()
        insightsStorageInstance.checkForGA()
        insightsStorageInstance.checkLinks()
      }
    }, 5000)
    historyStorage.on('init add undo redo', runChecksCallback)
    settingsStorage.state('pageTitleDisabled').onChange(runChecksCallback)
    settingsStorage.state('pageTitle').onChange(runChecksCallback)
    workspaceStorage.state('iframe').onChange(({ type }) => {
      if (type === 'loaded') {
        runChecksCallback()
      }
    })
    elementsStorage.on('elementsRenderDone', runChecksCallback)
  }
})
