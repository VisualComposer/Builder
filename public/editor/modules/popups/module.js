import { add, getStorage, getService } from 'vc-cake'
import { getPopupDataFromElement } from 'public/tools/popup'
import store from 'public/editor/stores/store'
import { popupAddInProgressSet } from 'public/editor/stores/popup/slice'

const elementsStorage = getStorage('elements')
const dataProcessor = getService('dataProcessor')
const cook = getService('cook')

const addPopupHtml = (id) => {
  store.dispatch(popupAddInProgressSet(true))
  const contentWindow = window.document.querySelector('.vcv-layout-iframe').contentWindow
  const documentBody = contentWindow.document.body
  const domId = `vcv-popup-${id}`

  if (documentBody.querySelector(`#${domId}`)) {
    store.dispatch(popupAddInProgressSet(false))
    return
  }

  dataProcessor.appAdminServerRequest({
    'vcv-action': 'popupBuilder:getData:adminNonce',
    'vcv-source-id': id
  }).then((requestData) => {
    if (requestData && typeof requestData === 'string') {
      const popupContainer = contentWindow.document.createElement('div')
      popupContainer.id = domId
      popupContainer.className = 'vcv-popup-container'
      popupContainer.innerHTML = requestData
      documentBody.appendChild(popupContainer)
    }
    store.dispatch(popupAddInProgressSet(false))
  }, (error) => {
    store.dispatch(popupAddInProgressSet(false))
    console.warn('Failed to get the popup HTML', error)
  })
}

add('popups', () => {
  elementsStorage.on('add', (elementData) => {
    const cookElement = cook.get(elementData)
    const ids = getPopupDataFromElement(cookElement)
    ids.forEach((id) => {
      addPopupHtml(id)
    })
  })

  elementsStorage.on('update', (id) => {
    const cookElement = cook.getById(id)
    if (!cookElement) {
      return
    }
    const ids = getPopupDataFromElement(cookElement)
    ids.forEach((id) => {
      addPopupHtml(id)
    })
  })
})
