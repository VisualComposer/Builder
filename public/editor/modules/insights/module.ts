import { getStorage, env, add } from 'vc-cake'
import { debounce } from 'lodash'
import InsightsChecks from './lib/InsightsChecks'
import store from 'public/editor/stores/store'
import { insightsReset } from 'public/editor/stores/insights/slice'

const historyStorage = getStorage('history')
const settingsStorage = getStorage('settings')
const workspaceStorage = getStorage('workspace')
const elementsStorage = getStorage('elements')

add('insights', () => {
  if (typeof (window as any).vcvRebuildPostSave !== 'undefined') {
    return
  }
  if (env('VCV_FT_INSIGHTS')) {
    const insightsChecksInstance = new InsightsChecks()
    const runChecksCallback = debounce(() => {
      const iframeDomNode = document.querySelector<HTMLIFrameElement>('.vcv-layout-iframe')
      if (!iframeDomNode || !iframeDomNode.contentWindow) {
        return // editor reload
      }
      // clear previous <Insights>
      store.dispatch(insightsReset())
      insightsChecksInstance.isImagesSizeLarge = false

      // Do all checks
      const isEmptyContent: boolean = insightsChecksInstance.checkForEmptyContent()
      if (!isEmptyContent) {
        insightsChecksInstance.checkForHeadings()
        insightsChecksInstance.checkForAlt()
        insightsChecksInstance.checkNameForLinks()
        insightsChecksInstance.checkForImagesSize()
        insightsChecksInstance.checkParagraphsLength()
        insightsChecksInstance.checkTitleLength()
        insightsChecksInstance.checkNoIndex()
        insightsChecksInstance.checkPostContentLength()
        insightsChecksInstance.checkForGA()
        insightsChecksInstance.checkLinks()
        insightsChecksInstance.contrast()
      }
    }, 5000)
    historyStorage.on('init add undo redo', runChecksCallback)
    settingsStorage.state('pageTitleDisabled').onChange(runChecksCallback)
    settingsStorage.state('pageTitle').onChange(runChecksCallback)
    workspaceStorage.state('iframe').onChange(({ type }: { type: string }) => {
      if (type === 'loaded') {
        runChecksCallback()
      }
    })
    elementsStorage.on('elementsRenderDone', runChecksCallback)
  }
})
