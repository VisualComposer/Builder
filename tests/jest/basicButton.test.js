/* global expect, it */
import vcCake from 'vc-cake'

import '../../public/variables'
import '../../public/editor/services/utils/service.js'
import '../../public/editor/services/dataManager/service.js'
import '../../public/editor/services/document/service.js'
import '../../public/editor/services/dataProcessor/service.js'
import '../../public/editor/services/hubElements/service.js'
import '../../public/editor/services/cook/service.js'
import '../../public/editor/services/modernAssetsStorage/service.js'
import '../../public/editor/services/api/service.js'
import '../../public/editor/services/wordpress-post-data/service'

import '../../public/config/wp-attributes'

// import '../../public/editor/stores/shortcodesAssets/storage'
import '../../public/editor/stores/elements/elementSettings'

import '../../public/editor/modules/elementLimit/module'

import '../../public/tools/updateHtmlWithServer'


import React from 'react'
import renderer from 'react-test-renderer'
// import './__mocks__/api'

import ButtonElement from '../../elements/basicButton/basicButton/component'


describe('Tests Basic Button element render', () => {
  test('Basic Button render', () => {
    const text = 'Hello jest from react'

    vcCake.env('platform', 'jest').start(()=>{})
    const tree = renderer
      .create(<ButtonElement atts={{ buttonText: text }} id='123456789' />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})

