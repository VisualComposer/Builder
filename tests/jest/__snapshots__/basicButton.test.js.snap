// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests Basic Button element render Basic Button render 1`] = `
<div
  className="vce-button--style-basic-container"
>
  <span
    className="vce-button--style-basic-wrapper vce"
    data-vce-do-apply="margin el-123456789"
    id="el-123456789"
  >
    <button
      className="vce-button--style-basic"
      data-vce-do-apply="padding border background  el-123456789"
    >
      Hello jest from react
    </button>
  </span>
</div>
`;
